File: scripts/download-stations.mjs
import { readFile, writeFile } from "node:fs/promises";

const wikidataSparqlUrl = "https://query.wikidata.org/sparql";
const overpassUrls = [
  "https://overpass-api.de/api/interpreter",
  "https://overpass.kumi.systems/api/interpreter"
];

const countries = [
  { name: "Albania", qid: "Q222", iso: "AL" },
  { name: "Austria", qid: "Q40", iso: "AT" },
  { name: "Belgium", qid: "Q31", iso: "BE" },
  { name: "Bosnia and Herzegovina", qid: "Q225", iso: "BA" },
  { name: "Bulgaria", qid: "Q219", iso: "BG" },
  { name: "Croatia", qid: "Q224", iso: "HR" },
  { name: "Czechia", qid: "Q213", iso: "CZ" },
  { name: "Denmark", qid: "Q35", iso: "DK" },
  { name: "Estonia", qid: "Q191", iso: "EE" },
  { name: "Finland", qid: "Q33", iso: "FI" },
  { name: "France", qid: "Q142", iso: "FR" },
  { name: "Germany", qid: "Q183", iso: "DE" },
  { name: "Greece", qid: "Q41", iso: "GR" },
  { name: "Hungary", qid: "Q28", iso: "HU" },
  { name: "Ireland", qid: "Q27", iso: "IE" },
  { name: "Italy", qid: "Q38", iso: "IT" },
  { name: "Kosovo", qid: "Q1246", iso: "XK" },
  { name: "Latvia", qid: "Q211", iso: "LV" },
  { name: "Lithuania", qid: "Q37", iso: "LT" },
  { name: "Liechtenstein", qid: "Q347", iso: "LI" },
  { name: "Luxembourg", qid: "Q32", iso: "LU" },
  { name: "Moldova", qid: "Q217", iso: "MD" },
  { name: "Montenegro", qid: "Q236", iso: "ME" },
  { name: "Netherlands", qid: "Q55", iso: "NL" },
  { name: "Norway", qid: "Q20", iso: "NO" },
  { name: "Poland", qid: "Q36", iso: "PL" },
  { name: "Portugal", qid: "Q45", iso: "PT" },
  { name: "Romania", qid: "Q218", iso: "RO" },
  { name: "Republic of Serbia", qid: "Q403", iso: "RS" },
  { name: "Slovakia", qid: "Q214", iso: "SK" },
  { name: "Slovenia", qid: "Q215", iso: "SI" },
  { name: "Spain", qid: "Q29", iso: "ES" },
  { name: "Sweden", qid: "Q34", iso: "SE" },
  { name: "Switzerland", qid: "Q39", iso: "CH" },
  { name: "Ukraine", qid: "Q212", iso: "UA" },
  { name: "United Kingdom", qid: "Q145", iso: "GB" }
];

// -------------------- helpers --------------------

function parsePoint(value) {
  const match = /^Point\(([-\d.]+) ([-\d.]+)\)$/.exec(value || "");
  if (!match) return null;
  return { lng: Number(match[1]), lat: Number(match[2]) };
}

function cleanStationName(name) {
  return String(name || "")
    .replace(/\s+/g, " ")
    .replace(/^station\s+/i, "")
    .replace(/\s+station$/i, "")
    .replace(/\s+railway station$/i, "")
    .trim();
}

function isKilometerMarkerName(name) {
  return /(^|[^a-z])km([^a-z]|$)|\bkilometr/i.test(name || "");
}

function isFiniteCoord(lat, lng) {
  return Number.isFinite(lat) && Number.isFinite(lng);
}

function stableId(text) {
  return String(text || "")
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
}

function distanceMeters(a, b) {
  const R = 6371000;
  const lat1 = a.lat * Math.PI / 180;
  const lat2 = b.lat * Math.PI / 180;
  const dLat = (b.lat - a.lat) * Math.PI / 180;
  const dLng = (b.lng - a.lng) * Math.PI / 180;

  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;

  return 2 * R * Math.atan2(Math.sqrt(h), Math.sqrt(1 - h));
}

function namesCompatible(a, b) {
  const x = stableId(a);
  const y = stableId(b);
  if (!x || !y) return false;
  if (x === y) return true;

  const [s, l] = x.length < y.length ? [x, y] : [y, x];
  return s.length >= 5 && l.includes(s);
}

function stationKey(station) {
  return `${stableId(station.name)}:${Math.round(station.lat * 10000)}:${Math.round(station.lng * 10000)}`;
}

// -------------------- wikidata --------------------

async function fetchJson(url, options = {}) {
  const res = await fetch(url, options);
  if (!res.ok) throw new Error(`${res.status} ${res.statusText}`);
  return res.json();
}

function wikidataEntityId(url) {
  return String(url || "").split("/").pop();
}

async function fetchWikidataStations(country) {
  const query = `
    SELECT ?station ?stationLabel ?coord (GROUP_CONCAT(DISTINCT ?operatorLabel; separator=", ") AS ?operators)
    WHERE {
      ?station wdt:P31 wd:Q55488;
               wdt:P17 wd:${country.qid};
               wdt:P5817 wd:Q55654238;
               wdt:P625 ?coord.
      OPTIONAL { ?station wdt:P137 ?operator. }
      SERVICE wikibase:label {
        bd:serviceParam wikibase:language "nl,en,[AUTO_LANGUAGE]".
      }
    }
    GROUP BY ?station ?stationLabel ?coord
  `;

  const url = `${wikidataSparqlUrl}?format=json&query=${encodeURIComponent(query)}`;

  const data = await fetchJson(url, {
    headers: {
      Accept: "application/sparql-results+json",
      "User-Agent": "TrackAndTide/1.0"
    }
  });

  return (data.results?.bindings || [])
    .map(b => {
      const p = parsePoint(b.coord?.value);
      return {
        id: b.station?.value,
        wikidataId: wikidataEntityId(b.station?.value),
        source: "wikidata",
        name: cleanStationName(b.stationLabel?.value),
        lat: p?.lat,
        lng: p?.lng,
        operators: b.operators?.value?.split(", ") || []
      };
    })
    .filter(s =>
      s.id &&
      s.name &&
      !isKilometerMarkerName(s.name) &&
      isFiniteCoord(s.lat, s.lng)
    );
}

// -------------------- OSM --------------------

function isUsableOsmStation(tags = {}) {
  if (!tags.name || tags.railway !== "station") return false;

  const bad = new Set(["abandoned", "construction", "disused", "historic", "razed", "proposed"]);
  const badTypes = new Set(["subway", "light_rail", "tram", "monorail"]);

  if (bad.has(tags.lifecycle)) return false;
  if (badTypes.has(tags.station)) return false;

  if (
    tags.disused === "yes" ||
    tags.abandoned === "yes" ||
    tags.construction === "yes" ||
    tags.proposed === "yes"
  ) return false;

  if (tags.tourism === "museum") return false;

  return true;
}

async function fetchOsmStations(country) {
  const query = `
    [out:json][timeout:180];
    area["ISO3166-1"="${country.iso}"][admin_level=2]->.a;
    (
      node["railway"="station"](area.a);
      way["railway"="station"](area.a);
      relation["railway"="station"](area.a);
    );
    out center tags;
  `;

  let data;

  for (const url of overpassUrls) {
    try {
      data = await fetchJson(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          "User-Agent": "TrackAndTide/1.0"
        },
        body: new URLSearchParams({ data: query })
      });
      break;
    } catch (e) {
      console.warn(`OSM failed ${url}: ${e.message}`);
    }
  }

  if (!data) throw new Error("All OSM endpoints failed");

  return (data.elements || [])
    .filter(e => isUsableOsmStation(e.tags))
    .map(e => {
      const t = e.tags || {};
      const lat = e.lat ?? e.center?.lat;
      const lng = e.lon ?? e.center?.lon;

      return {
        id: `osm:${e.type}/${e.id}`,
        source: "osm",
        wikidataId: t.wikidata || "",
        name: cleanStationName(t.name),
        lat: Number(lat),
        lng: Number(lng),
        operators: t.operator ? t.operator.split(";") : []
      };
    })
    .filter(s =>
      s.name &&
      !isKilometerMarkerName(s.name) &&
      isFiniteCoord(s.lat, s.lng)
    );
}

// -------------------- merge --------------------

function mergeStations(wikidataStations, osmStations) {
  const merged = [];
  const seen = new Set();

  const wikidataById = new Map(
    wikidataStations
      .filter(s => s.wikidataId)
      .map(s => [s.wikidataId, s])
  );

  // 1. add Wikidata first
  for (const w of wikidataStations) {
    seen.add(stationKey(w));
    merged.push(w);
  }

  // 2. add OSM with rules
  for (const o of osmStations) {
    const key = stationKey(o);
    if (seen.has(key)) continue;

    // rule 1: OSM has wikidata → skip if exists
    if (o.wikidataId && wikidataById.has(o.wikidataId)) {
      continue;
    }

    // rule 2: very close station → keep Wikidata only (<50m)
    const near = merged.some(m => distanceMeters(o, m) <= 50);
    if (near) continue;

    seen.add(key);
    merged.push(o);
  }

  return merged.sort((a, b) => a.name.localeCompare(b.name));
}

// -------------------- pipeline --------------------

async function downloadCountry(country) {
  console.log(`Downloading ${country.name}...`);

  const [w, o] = await Promise.allSettled([
    fetchWikidataStations(country),
    fetchOsmStations(country)
  ]);

  const wikidataStations = w.status === "fulfilled" ? w.value : [];
  const osmStations = o.status === "fulfilled" ? o.value : [];

  const stations = mergeStations(wikidataStations, osmStations);

  console.log(`  ${stations.length} stations`);

  return {
    country: country.name,
    qid: country.qid,
    iso: country.iso,
    counts: {
      total: stations.length,
      wikidata: stations.filter(s => s.source === "wikidata").length,
      osm: stations.filter(s => s.source === "osm").length
    },
    stations
  };
}

// -------------------- CLI --------------------

const requested = process.argv.slice(2).filter(a => !a.startsWith("--"));

const selected = requested.length
  ? countries.filter(c =>
      requested.includes(c.name) || requested.includes(c.iso)
    )
  : countries;

if (!selected.length) {
  console.error("No countries matched");
  process.exit(1);
}

let output = {
  generatedAt: new Date().toISOString(),
  sources: { wikidata: wikidataSparqlUrl, osm: overpassUrls },
  countries: {}
};

try {
  if (requested.length) {
    output = JSON.parse(await readFile("stations.json", "utf8"));
  }
} catch {}

for (const c of selected) {
  output.countries[c.name] = await downloadCountry(c);
}

await writeFile("stations.json", JSON.stringify(output, null, 2));
console.log("Saved stations.json");

File: scripts/download-stations.py
"""
Track & Tide — Station Downloader (Python)
Downloads train stations from Wikidata for all European countries.
Uses requests with proper timeout handling.
"""
import json, sys, os, time, requests
from math import radians, cos, sin, asin, sqrt

EXISTING_FILE = os.path.join(os.path.dirname(__file__), '..', 'stations.json')
OUTPUT_FILE = EXISTING_FILE
STATE_FILE = os.path.join(os.path.dirname(__file__), '..', 'station-download-state.json')

EUROPEAN_COUNTRIES = [
    ('Albania', 'AL', 'Q222'),
    ('Andorra', 'AD', 'Q228'),
    ('Austria', 'AT', 'Q40'),
    ('Belarus', 'BY', 'Q184'),
    ('Belgium', 'BE', 'Q31'),
    ('Bosnia and Herzegovina', 'BA', 'Q225'),
    ('Bulgaria', 'BG', 'Q219'),
    ('Croatia', 'HR', 'Q224'),
    ('Cyprus', 'CY', 'Q229'),
    ('Czechia', 'CZ', 'Q213'),
    ('Denmark', 'DK', 'Q35'),
    ('Estonia', 'EE', 'Q191'),
    ('Finland', 'FI', 'Q33'),
    ('France', 'FR', 'Q142'),
    ('Germany', 'DE', 'Q183'),
    ('Greece', 'GR', 'Q41'),
    ('Hungary', 'HU', 'Q28'),
    ('Iceland', 'IS', 'Q189'),
    ('Ireland', 'IE', 'Q27'),
    ('Italy', 'IT', 'Q38'),
    ('Kosovo', 'XK', 'Q1246'),
    ('Latvia', 'LV', 'Q211'),
    ('Liechtenstein', 'LI', 'Q347'),
    ('Lithuania', 'LT', 'Q37'),
    ('Luxembourg', 'LU', 'Q32'),
    ('Malta', 'MT', 'Q233'),
    ('Moldova', 'MD', 'Q217'),
    ('Monaco', 'MC', 'Q235'),
    ('Montenegro', 'ME', 'Q236'),
    ('Netherlands', 'NL', 'Q55'),
    ('North Macedonia', 'MK', 'Q221'),
    ('Norway', 'NO', 'Q20'),
    ('Poland', 'PL', 'Q36'),
    ('Portugal', 'PT', 'Q45'),
    ('Romania', 'RO', 'Q218'),
    ('San Marino', 'SM', 'Q238'),
    ('Serbia', 'RS', 'Q403'),
    ('Slovakia', 'SK', 'Q214'),
    ('Slovenia', 'SI', 'Q215'),
    ('Spain', 'ES', 'Q29'),
    ('Sweden', 'SE', 'Q34'),
    ('Switzerland', 'CH', 'Q39'),
    ('Turkey', 'TR', 'Q43'),
    ('Ukraine', 'UA', 'Q212'),
    ('United Kingdom', 'GB', 'Q145'),
    ('Vatican City', 'VA', 'Q237'),
]

DEDUP_THRESHOLD_KM = 0.5
TIMEOUT = 60  # seconds per SPARQL query

def haversine_km(lat1, lng1, lat2, lng2):
    R = 6371
    dlat = radians(lat2 - lat1)
    dlng = radians(lng2 - lng1)
    a = sin(dlat/2)**2 + cos(radians(lat1)) * cos(radians(lat2)) * sin(dlng/2)**2
    return R * 2 * asin(sqrt(a))

def clean_name(name):
    import re
    name = str(name or '')
    name = re.sub(r'\s+', ' ', name)
    name = re.sub(r'^Bahnhof\s+', '', name, flags=re.IGNORECASE)
    name = re.sub(r"^(Gare de|Gare du|Gare d'|Stazione di|Stazione|Estación de|Estación|Estação de|Estação|Dworzec)\s+", '', name, flags=re.IGNORECASE)
    name = re.sub(r'\s+(railway station|railway|train station|central station|bus station|metro station|airport|aeroport|aéroport|flughafen|hauptbahnhof|hbf|bahnhof|station|gare|stazione|estación|estação|vasútállomás|pályaudvar|állomás)$', '', name, flags=re.IGNORECASE)
    return name.strip()

def make_id(name, iso):
    import re, unicodedata
    clean = clean_name(name).lower()
    clean = unicodedata.normalize('NFD', clean)
    clean = ''.join(c for c in clean if not unicodedata.combining(c))
    clean = re.sub(r'[^a-z0-9]+', '-', clean).strip('-')
    return f"{clean}-{iso.lower()}"

def fetch_wikidata(country_name, qid):
    # Strategy 1: Fast query with label service (GET)
    query = f"""
SELECT DISTINCT ?station ?stationLabel ?wdId ?lat ?lon WHERE {{
  ?station wdt:P17 wd:{qid}.
  ?station wdt:P31 ?type.
  VALUES ?type {{
    wd:Q55488    # railway station
    wd:Q55489    # railway stop (halt)
    wd:Q928830   # metro station
    wd:Q12046342 # S-Bahn station
    wd:Q2175765  # underground station
    wd:Q55491    # railway junction
    wd:Q548662   # request stop / halt
  }}
  ?station wdt:P625 ?coord.
  BIND(geof:latitude(?coord) AS ?lat)
  BIND(geof:longitude(?coord) AS ?lon)
  BIND(STRAFTER(STR(?station), "http://www.wikidata.org/entity/") AS ?wdId)
  SERVICE wikibase:label {{ bd:serviceParam wikibase:language "en,de,fr,it,es,nl,pl,cs,sv,da,fi,el,hu,ro,bg,sk,sl,hr,sr,lt,lv,et,uk,no". }}
}}
LIMIT 30000
"""
    try:
        resp = requests.get('https://query.wikidata.org/sparql',
                          params={'format': 'json', 'query': query},
                          headers={'User-Agent': 'TrackAndTide/1.0'},
                          timeout=TIMEOUT)
        resp.raise_for_status()
        try:
            data = resp.json()
        except json.JSONDecodeError:
            data = json.loads(resp.text, strict=False)
        bindings = data.get('results', {}).get('bindings', [])
        if bindings:
            return parse_bindings(bindings)
    except Exception as e:
        print(f"    Wikidata GET failed: {e}, trying paginated fallback...")
    
    # Strategy 2: Fallback with pagination, no label service
    return fetch_wikidata_paginated(country_name, qid)

def parse_bindings(bindings):
    stations = []
    for b in bindings:
        try:
            lat = float(b['lat']['value'])
            lng = float(b['lon']['value'])
            wdid = b.get('wdId', {}).get('value', '')
            name = b.get('stationLabel', {}).get('value', '')
            if not name:
                name = b.get('station', {}).get('value', '').split('/')[-1].replace('_', ' ')
            if lat and lng and name:
                stations.append({
                    'name': name, 'lat': lat, 'lng': lng,
                    'wikidataId': wdid, 'source': 'wikidata',
                    'url': f"https://www.wikidata.org/wiki/{wdid}" if wdid else '',
                })
        except (ValueError, KeyError):
            pass
    return stations

def fetch_wikidata_paginated(country_name, qid):
    all_stations = []
    offset = 0
    batch_size = 5000
    
    while offset < 50000:
        query = f"""
SELECT DISTINCT ?station ?stationLabel ?wdId ?lat ?lon WHERE {{
  ?station wdt:P17 wd:{qid}.
  ?station wdt:P31 ?type.
  VALUES ?type {{
    wd:Q55488 wd:Q55489 wd:Q928830 wd:Q12046342 wd:Q2175765 wd:Q55491 wd:Q548662
  }}
  ?station wdt:P625 ?coord.
  ?station rdfs:label ?stationLabel.
  FILTER(LANG(?stationLabel) = "en" || LANG(?stationLabel) = "de" || LANG(?stationLabel) = "fr" 
      || LANG(?stationLabel) = "it" || LANG(?stationLabel) = "es" || LANG(?stationLabel) = "nl"
      || LANG(?stationLabel) = "pl" || LANG(?stationLabel) = "cs")
  BIND(geof:latitude(?coord) AS ?lat)
  BIND(geof:longitude(?coord) AS ?lon)
  BIND(STRAFTER(STR(?station), "http://www.wikidata.org/entity/") AS ?wdId)
}}
LIMIT {batch_size} OFFSET {offset}
"""
        try:
            resp = requests.get('https://query.wikidata.org/sparql',
                              params={'format': 'json', 'query': query},
                              headers={'User-Agent': 'TrackAndTide/1.0'},
                              timeout=TIMEOUT)
            resp.raise_for_status()
            try:
                data = resp.json()
            except json.JSONDecodeError:
                data = json.loads(resp.text, strict=False)
            bindings = data.get('results', {}).get('bindings', [])
            if not bindings:
                break
            all_stations.extend(parse_bindings(bindings))
            if len(bindings) < batch_size:
                break
            offset += batch_size
        except Exception as e:
            print(f"    Wikidata fallback error at offset {offset}: {e}")
            break
    
    return all_stations

def deduplicate(stations):
    deduped = []
    for s in stations:
        dup = False
        for existing in deduped:
            if haversine_km(s['lat'], s['lng'], existing['lat'], existing['lng']) < DEDUP_THRESHOLD_KM:
                dup = True
                if s.get('operators') and not existing.get('operators'):
                    existing['operators'] = s['operators']
                if s['source'] == 'wikidata' and existing['source'] == 'osm':
                    existing['wikidataId'] = s.get('wikidataId', '')
                    existing['url'] = s.get('url', '')
                    existing['source'] = 'wikidata'
                break
        if not dup:
            deduped.append(s)
    deduped.sort(key=lambda x: x['name'])
    return deduped

def merge_osm(new_stations, existing_country):
    if not existing_country or 'stations' not in existing_country:
        return new_stations, 0
    osm_preserved = 0
    for old in existing_country['stations']:
        if old.get('source') != 'osm':
            continue
        if not old.get('lat') or not old.get('lng'):
            continue
        dup = False
        for ns in new_stations:
            if ns.get('source') == 'osm':
                continue
            if haversine_km(old['lat'], old['lng'], ns['lat'], ns['lng']) < DEDUP_THRESHOLD_KM:
                dup = True
                if old.get('operators') and not ns.get('operators'):
                    ns['operators'] = old['operators']
                if old.get('limited_use'):
                    ns['limited_use'] = True
                break
        if not dup:
            new_stations.append({
                'id': old.get('id', ''),
                'url': old.get('url', ''),
                'wikidataId': old.get('wikidataId', ''),
                'source': 'osm',
                'name': old['name'],
                'lat': old['lat'],
                'lng': old['lng'],
                'operators': old.get('operators', []),
                'limited_use': old.get('limited_use', False),
            })
            osm_preserved += 1
    return new_stations, osm_preserved

def main():
    # Parse args
    target_names = None
    args = sys.argv[1:]
    if '--country' in args:
        idx = args.index('--country')
        target_names = set()
        for a in args[idx+1:]:
            if a.startswith('--'):
                break
            target_names.add(a)
    
    # Load existing
    existing = {'countries': {}}
    try:
        with open(EXISTING_FILE, 'r', encoding='utf-8') as f:
            existing = json.load(f)
        print(f"Loaded existing data with {len(existing.get('countries', {}))} countries")
    except:
        print("No existing data found")
    
    # Filter countries
    target = EUROPEAN_COUNTRIES
    if target_names:
        target = [c for c in EUROPEAN_COUNTRIES if c[0] in target_names]
    
    print(f"Target: {', '.join(c[0] for c in target)}")
    print()
    
    result = {}
    total_wd = 0
    total_osm = 0
    grand_total = 0
    
    for i, (name, iso, qid) in enumerate(target):
        print(f"[{i+1}/{len(target)}] {name} ({iso})")
        
        wd = fetch_wikidata(name, qid)
        
        if wd is None:
            # Failed, use existing
            existing_country = existing.get('countries', {}).get(name)
            if existing_country:
                result[name] = existing_country
                cnt = existing_country.get('counts', {}).get('total', 0)
                print(f"  Wikidata failed, preserving {cnt} existing stations")
                grand_total += cnt
            else:
                result[name] = {
                    'country': name, 'qid': qid, 'iso': iso,
                    'counts': {'total': 0, 'wikidata': 0, 'osm': 0},
                    'stations': [],
                }
            print()
            continue
        
        print(f"    Wikidata returned {len(wd)} results")
        
        stations = []
        for s in wd:
            stations.append({
                'id': make_id(s['name'], iso),
                'url': s['url'],
                'wikidataId': s['wikidataId'],
                'source': s['source'],
                'name': s['name'],
                'lat': s['lat'],
                'lng': s['lng'],
                'operators': [],
            })
        
        wd_before = len(stations)
        stations = deduplicate(stations)
        wd_after = len(stations)
        
        existing_country = existing.get('countries', {}).get(name)
        stations, osm_preserved = merge_osm(stations, existing_country)
        
        stations = deduplicate(stations)
        
        total_wd += len(wd)
        total_osm += osm_preserved
        grand_total += len(stations)
        
        result[name] = {
            'country': name, 'qid': qid, 'iso': iso,
            'counts': {'total': len(stations), 'wikidata': len(wd), 'osm': osm_preserved},
            'stations': stations,
        }
        
        print(f"  → {len(stations)} unique ({wd_before} WD raw, {wd_after} WD dedup, {osm_preserved} OSM preserved)")
        print()
        
        # Save intermediate after each country
        output = {
            'generatedAt': time.strftime('%Y-%m-%dT%H:%M:%S.000Z', time.gmtime()),
            'sources': {
                'wikidata': 'https://query.wikidata.org/sparql',
                'osm': existing.get('sources', {}).get('osm', ['https://overpass-api.de/api/interpreter']),
            },
            'countries': result,
        }
        # Also preserve unprocessed existing countries
        for cn, cd in existing.get('countries', {}).items():
            if cn not in result:
                result[cn] = cd
        
        output['countries'] = result
        with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
            json.dump(output, f, ensure_ascii=False)
    
    # Final summary
    print(f"\n{'═'*50}")
    print(f"Download complete!")
    print(f"Countries processed: {len(target)}")
    print(f"Total Wikidata fetched: {total_wd}")
    print(f"Total OSM preserved: {total_osm}")
    print(f"Total unique stations: {grand_total}")
    print(f"Output: {OUTPUT_FILE}")
    print(f"{'═'*50}")

if __name__ == '__main__':
    main()
